ecosystem/sep-0040.md

```diff
@@ -0,0 +1,198 @@
+
+**Backward Compatibility:** Existing NFT platforms essentially default to `type: ipfshash`. While this standard aims to free up the ecosystem from dependence on IPFS, it should also work with existing conventions that developed organically. To do so, **if you see an account with a lone `ipfshash` data entry, you may treat it as equivalent to an NFT account with `type: ipfshash`**.
+
+##### IPFS Storage
+If the `type` is `ipfshash`, the `ipfshash` key must point to an [IPFS Content IDentifier](https://docs.ipfs.io/concepts/content-addressing/) corresponding to the location of the metadata.
+
+If you are the issuer, make sure you understand the implications of using IPFS. You must use an IPFS lookup service to actually resolve to the metadata. For example, [this URL](https://cloudflare-ipfs.com/ipfs/QmaydGEshbL7HMQkD6FZPcMeEt6o15aC4p4P82VxQ7amjy) resolves to the Meridian 2021 Hat NFT via the Cloudflare IPFS service. If you're the issuer, you must also ensure that the resource is addressable and accessible on the IPFS network by running your own IPFS node or entrusting a provider to do so.
+
+##### URL Storage
+If the `type` is `uri`, the corresponding `uri` key must point to a [valid URI](https://datatracker.ietf.org/doc/html/rfc3986) containing the NFT's metadata. 
+
+_Note_: While you could technically describe metadata living on the IPFS with a URL by pointing to a specific node or a URI prefixing the CID with `ipfs://`, we recommend describing _only_ its CID to leave lookups to the users.
+
+#### Standard Metadata Specification
+This section describes the metadata itself rather than how to find it. 
+
+The metadata must be valid JSON and must contain the following **required** fields:
+
+| Field       | Description |
+| ----------- | ----------- |
+| `name` | a short name or title for the NFT |
+| `description` | a longer, more-detailed description of the NFT |
+| `type` | a [Media Type](https://www.iana.org/assignments/media-types/media-types.xhtml) describing the NFT's primary format (pointed to by `url`) |
+| `uri` **or** `url`  | a [URI](https://datatracker.ietf.org/doc/html/rfc3986) pointing to the actual NFT contents |
+| `issuer` | a Stellar public key ([`G...`](https://stellar.org/protocol/sep-23)) corresponding to the account which issued this NFT |
+| `code` | the NFT's asset code |
+
+We allow either `uri` or `url` to maintain compatibility with existing ecosystem conventions. If _both_ are present, prefer the `uri`.
+
+For example,
+
+```json
+{
+  "name": "A demonstration of NFT metadata",
+  "description": "This is a description of the cool NFT used in this demo.",
+  "type": "image/png",
+  "url": "https://cloudflare-ipfs.com/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/I/m/RickAstleyNeverGonnaGiveYouUp7InchSingleCover.jpg",
+  "issuer": "GALAXYVOIDAOPZTDLHILAJQKCVVFMD4IKLXLSZV5YHO7VY74IWZILUTO",
+  "code": "DEMOASSET"
+}
+```
+
```

Comment:
Why not support the concept of a single NFT pointing to multiple pieces of content as an array of content items? This would allow a single NFT to point to a metadata file that in turn has pointers to multiple resources (e.g. a video, a thumbnail of the video, a certificate document).

This could be represented by a `data` top-level property that always contains an Array of at least one object. each object in the array could have additional attributes.

For example,

```json
{
  "name": "A demonstration of NFT metadata",
  "description": "This is a description of the cool NFT used in this demo.",
  "data": [
    {
     "description": " This is a cool image",
     "type": "image/png",
     "url": "https://filename.jpg",
     "hash": "deadbeef01",
     "hashType": "sha2-256",
    },
    {
     "description": " This is a cool video",
     "type": "video/mpeg",
     "url": "https://filename.mp3",
     "hash": "deadbeef02",
     "hashType": "sha2-512",
    }
  ],
  "issuer": "GALAXYVOIDAOPZTDLHILAJQKCVVFMD4IKLXLSZV5YHO7VY74IWZILUTO",
  "code": "DEMOASSET"
}
```

This would allow new types of multi-part assets to be referenced in a single NFT.

Something that will have to be addressed in any case is also how to canonicalize the contents of the metadata file so that it can be hashed consistently by both the creator and the verifier. JSON objects are not guaranteed to have their contents in any particular order. A standard method for ordering the metadata file so that it can be consistently hashed by all parties must be defined. Otherwise, it may not be possible to verify the on-chain metadata content hash.

See for example:

https://github.com/SkeLLLa/node-object-hash#readme

Reply:
The multi-part suggestion is really nice, thanks! We want to maximize compatibility, though, so we may have to resort to "either a `data` array or the same fields on the root object" which I hate but aligns with existing NFTs.

Regarding JSON ordering: The hash should be computed on the raw metadata file rather than the "loaded" JSON blob in order to avoid this. It's an important subtlety, so I'll note it.

Reply:
We could support the URL/ipfshash being to a directory, and inside the directory each asset code could get its own .json, so that an account can issue multiple asset codes but share the issuing account. This should work with most URL protocols.